src/hotspot/share/opto/mulnode.cpp

```diff
@@ -270,11 +270,20 @@ Node *MulINode::Ideal(PhaseGVN *phase, bool can_reshape) {
     unsigned int bit2 = abs_con - bit1;
     bit2 = bit2 & (0 - bit2);          // Extract 2nd bit
     if (bit2 + bit1 == abs_con) {    // Found all bits in con?
+      if (!phase->C->post_loop_opts_phase()) {
+        // Defer this because it breaks loop range check hoisting
+        phase->C->record_for_post_loop_opts_igvn(this);
+        return NULL;
+      }
```

Comment:
This defers the whole idealisation, should we only skip this particular transformation instead? Also should this be applied to `MulLNode`, too? Thanks.

Reply:
I think this only defers the transformation if scale value is a constant and has exactly 2 bits set in binary. Could you elaborate how to make it more particular?

AFAIK, all normal Java array accesses are using 32-bit indices. When running on a 64-bit platform, we use a `ConvI2L` in element address computing but it's done after whole index expression computing. Is there any special array accesses that may use `MulL`?

Reply:
I mean this stops the whole idealisation as soon as the constant has exactly 2 bits set, I think we should still try other transformations in `MulNode::Ideal` in those cases.

IIRC, memory segment accesses use long arithmetic, so they need this changes, too.

Thanks.

Reply:
Thanks for your suggestions. I will update the code later after some tests.